### PR TITLE
HPCC-17122 Better tracking of scoped context

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser please review.  tested by running regression suite.  Useful for tying down what was causing John's problem.

**Review by shamser — APPROVED** [review]

File: ecl/hql/hqltrans.cpp [hunk 1]
```diff
@@ -4234,15 +4235,30 @@ IHqlExpression * ScopedTransformer::getEvaluateScope(IHqlExpression * scope)
...
+    {
+        printf("----- nesting -----\n");
+        ForEachItemIn(i, scopeStack)
+        {
+            ScopeInfo & cur = scopeStack.item(i);
+            StringBuffer temp;
+            toUserECL(temp, cur.context, false);
+            printf("%p: %s", cur.dataset, temp.str());
+        }
+        maxDepth = scopeStack.ordinality();
```
Comment (shamser) [inline ecl/hql/hqltrans.cpp]: @ghalliday Why is maxDepth being updated to the depth the current scope?

**Review by shamser — COMMENTED** [review]
@ghalliday MaxDepth is being updated in pushScope.  Not sure why.

File: ecl/hql/hqltrans.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqltrans.cpp]: @shamser so they I only get reports when the stack gets even deeper.  I'm interested in what the mostly deeply nested instance is - so once I have an example at one level, I only want to know if there is a deeper once.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed ok.

**Pull request merged**